### ISSUE TITLE
Trouble with Strings larger than 64k

### ISSUE THREAD

**Issue by domguinard** [vznet/mongo-jackson-mapper#16]
First let me tell you how much we like the jackson mongo mapper, awesome piece of work! Big up!

Now, is there any plan to update the version of bson4jackson you are using? Version 1.3 has some important issues, one of which is that you can't insert strings bigger than 64k, this will simply crash the converter and the subsequent requests (https://github.com/michel-kraemer/bson4jackson/issues/18).

As a consequence the mongo-jackson-mapper does not support strings that are bigger than 64k. Bson4jackson is now at version 2.0.0 and this bug was solved (amongst others).

Any plans of upgrading the version?

**Comment by domguinard** [comment]
This is the type of exceptions we are getting: 

[...] threw exception
java.lang.IllegalStateException: Current state = CODING_END, new state = CODING
    at java.nio.charset.CharsetDecoder.throwIllegalStateException(CharsetDecoder.java:951)
    at java.nio.charset.CharsetDecoder.decode(CharsetDecoder.java:537)
    at de.undercouch.bson4jackson.io.LittleEndianInputStream.readUTF(LittleEndianInputStream.java:311)
    at de.undercouch.bson4jackson.io.LittleEndianInputStream.readUTF(LittleEndianInputStream.java:264)
    at de.undercouch.bson4jackson.BsonParser.readString(BsonParser.java:506)
    at de.undercouch.bson4jackson.BsonParser.nextToken(BsonParser.java:228)
    at org.codehaus.jackson.map.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:693)
    at org.codehaus.jackson.map.deser.BeanDeserializer.deserialize(BeanDeserializer.java:580)
    at org.codehaus.jackson.map.ObjectMapper._readValue(ObjectMapper.java:2695)
    at org.codehaus.jackson.map.ObjectMapper.readValue(ObjectMapper.java:1308)
    at net.vz.mongodb.jackson.internal.stream.JacksonDBDecoder.decode(JacksonDBDecoder.java:62)
    at com.mongodb.Response.<init>(Response.java:66)
    at com.mongodb.DBPort.go(DBPort.java:128)
    at com.mongodb.DBPort.call(DBPort.java:79)
    at com.mongodb.DBTCPConnector.call(DBTCPConnector.java:218)
    at com.mongodb.DBApiLayer$MyCollection.__find(DBApiLayer.java:305)
    at com.mongodb.DBCursor._check(DBCursor.java:369)
    at com.mongodb.DBCursor._hasNext(DBCursor.java:498)
    at com.mongodb.DBCursor.hasNext(DBCursor.java:523)
    at net.vz.mongodb.jackson.DBCursor.hasNext(DBCursor.java:315)
    at net.vz.mongodb.jackson.JacksonDBCollection.findOne(JacksonDBCollection.java:890) [...]

Thanks a lot!

**Comment by jroper** [comment]
bson4jackson 2 is only compatible with Jackson 2, which is a massive backwards incompatible upgrade from Jackson 1.9 (the package names all changed from org.codehaus.jackson to com.fasterxml.jackson).  Work upgrading to Jackson 2 is well underway, but the best solution to this in the short term will be backporting the fix in bson4jackson to 1.x, because I'm sure there will be other people that want to use bson4jackson with Jackson 1.x.  If you have the time, you could backport this fix yourself, and issue a pull request, otherwise I'll get around to it at some point when I have time.

**Comment by juanformoso** [comment]
Hi! We are having this exact issue. I see the last message was 3 months ago, any news on a fix or workaround?

**Comment by jroper** [comment]
This project has been moved to:

http://mongojack.org

If this is still an issue, please reraise the issue there.

**Comment by klinchaev** [comment]
Here is a simple patch to fix version 1.3.0 (you should edit 1 line in 1 file and rebuild bson4jackson).

``` diff
diff --git a/src/main/java/de/undercouch/bson4jackson/io/LittleEndianInputStream.java b/src/main/java/de/undercouch/bson4jackson/io/LittleEndianInputStream.java                
index 0454758..12fff29 100644                                                                                                                                                   
--- a/src/main/java/de/undercouch/bson4jackson/io/LittleEndianInputStream.java                                                                                                  
+++ b/src/main/java/de/undercouch/bson4jackson/io/LittleEndianInputStream.java
@@ -280,7 +280,8 @@ public class LittleEndianInputStream extends FilterInputStream implements DataIn
                ByteBuffer utf8buf = _staticBuffers.byteBuffer(UTF8_BUFFER, 1024 * 8);
                byte[] rawUtf8Buf = utf8buf.array();

-               CharsetDecoder dec = getUTF8Decoder();
+               // PATCHED CharsetDecoder dec = getUTF8Decoder();
+               CharsetDecoder dec = Charset.forName("UTF-8").newDecoder();
                int expectedLen = (len > 0 ? (int)(dec.averageCharsPerByte() * len) + 1 : 1024);
                CharBuffer cb = _staticBuffers.charBuffer(UTF8_BUFFER, expectedLen);
                try {
```
